### PR TITLE
Apply block-level bucketing

### PR DESCRIPTION
**Pull request by ruisizhang123**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #13
* #12
* #11
* #10

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -1608,7 +1621,7 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
                     self.nodes, all_gather_before_last_wait=False
                 )
                 self.nodes = fsdp_reorder.reorder_reduce_scatter(self.nodes, front_node)
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py:1623]: in previous PRs, `front_node` is used before defined.

File: torch/_inductor/simple_fsdp/bucket_helpers.py [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
+from typing import List, Tuple, Union
 
 import torch
 import torch.distributed as dist
-from .. import dependencies, ir, scheduler
+
+from .. import ir, scheduler
 from ..virtualized import V
 
+
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/bucket_helpers.py:10]: it seems you only applied linting to this PR, not the previous one

File: torch/_inductor/simple_fsdp/fsdp_bucket.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
+import math
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/fsdp_bucket.py:1]: call this file `transformer_block_bucketing.py`
